### DISCUSSION TITLE
Generate DAG's `doc_md` automatically  based on module/dag/task docstrings ?

### DISCUSSION THREAD

**Discussion by pedro-cf** [apache/airflow#39754] [Q&A] [answered]
Greetings, would it be possible to generate a DAG's `doc_md` automatically  based on module/dag/task docstrings ?

**Comment by Taragolis** [comment, marked as answer]
Automatically it could only set for task-flow / dag decorators

```python
from __future__ import annotations

from datetime import datetime, timezone
from airflow.decorators import task
from airflow.models.dag import dag


@dag(
    schedule=None,
    start_date=datetime(2024, 2, 29, tzinfo=timezone.utc),
    catchup=False,
    tags=["discussion", "39754", "implicit doc_md"],
)
def discussion_39754_example():
    """
    DAG which do something awesome.

    ![image](https://imgs.xkcd.com/comics/dependency_2x.png)
    """

    @task
    def do_nothing() -> str:
        """
        Task that does nothing.

        ![image](https://imgs.xkcd.com/comics/standards_2x.png)
        """
        return "Foo Bar"

    do_nothing()


_ = discussion_39754_example()

```

### DAG Documentation

![image](https://github.com/apache/airflow/assets/3998685/57e73f58-ac1c-4cce-87ec-8bf32f8e8163)


### Task Documentation

> [!IMPORTANT]  
> Task documentation doesn't show in grid view, see: https://github.com/apache/airflow/issues/39676

![image](https://github.com/apache/airflow/assets/3998685/ddb6ac0d-64de-481f-ac81-d748aae77f7e)

**Reply by pedro-cf** [reply]
Incredible, thank you!

**Reply by pedro-cf** [reply]
Side question:  

Why `_ = discussion_39754_example()`over just `discussion_39754_example()`

**Reply by Taragolis** [reply]
> Why _ = discussion_39754_example()over just discussion_39754_example()

Doesn't matter, it is come from the past version of Airflow, with modern one there is no differences

**Comment by pedro-cf** [comment]
@Taragolis Actually what would be more interesting is a generated DAG doc for the whole dag information with several sections:

-  overview  (docstring of DAG)
- parameters
- Tasks listed with dependency ordering and respective description (docstring of Task)


right now I just use: `doc_md=__doc__` and place the docs at the top of the dag file.

example:

-----

# DAG ID: `stellar_saga`: Stellar Saga Chronicles

## Overview
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed et ligula in ante malesuada congue. Duis luctus felis ut neque condimentum, eget malesuada ex dapibus. Integer nec nulla ac leo eleifend accumsan. Donec interdum tincidunt sapien, sed tincidunt est feugiat ut. Aliquam lacinia euismod orci ut dictum. Nulla a tortor at sapien ultricies egestas. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. 

## Parameters
- `stellar_date`: `(str)` Nullam nec justo nec sem finibus faucibus in at quam. Sed ut nisl sed dui gravida venenatis vel a odio. Donec in dui a nulla mollis ultrices.
- `cosmic_end_date`: `(str)` Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Sed sollicitudin nulla nec ultrices convallis. 
- `galactic_satellite`: `(str)` Fusce rhoncus enim nec urna placerat, sed dignissim velit feugiat. In sagittis pretium felis, sit amet fermentum nulla laoreet in. 
- `interstellar_geojson`: `(str)` Maecenas rutrum dolor et nulla lobortis, a molestie elit pellentesque. Morbi eu urna eget eros luctus posuere.
- `maximum_nebula_coverage`: `(int)` Maecenas a metus quis urna lobortis tristique. Sed fermentum, tortor at feugiat mattis, nunc augue ullamcorper eros, a viverra nibh ex id quam.
- `astral_crs`: `(str, optional)` Aenean vel leo sit amet sem lobortis consequat eget a sem.
- `constellation_indices`: `(list of str)` Suspendisse sit amet quam nec leo tincidunt ultrices a sit amet est. Nullam nec turpis a dolor sollicitudin sagittis.

## Tasks

### 1. `explore_asteroid_field`
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed et ligula in ante malesuada congue. Duis luctus felis ut neque condimentum, eget malesuada ex dapibus.

### 2. `navigate_spacecraft`
Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Sed sollicitudin nulla nec ultrices convallis. 

### 3. `gather_starlight_data`
Fusce rhoncus enim nec urna placerat, sed dignissim velit feugiat.

### 4. `process_galactic_images`
Maecenas rutrum dolor et nulla lobortis, a molestie elit pellentesque.

### 5. `analyze_cosmic_radiation`
Maecenas a metus quis urna lobortis tristique. Sed fermentum, tortor at feugiat mattis, nunc augue ullamcorper eros.

### 6. `stargate_transmission`
Aenean vel leo sit amet sem lobortis consequat eget a sem.

### 7. `catalog_stellar_phenomena`
Suspendisse sit amet quam nec leo tincidunt ultrices a sit amet est.

### 8. `celestial_navigation`
Morbi consectetur nisl vitae erat ultrices, quis fermentum velit congue.

### 9. `interstellar_communication`
Fusce lobortis arcu et lectus commodo, id suscipit ipsum eleifend.

### 10. `expedition_cleanup`
Sed vestibulum urna vitae sapien vehicula, vitae varius metus congue.

### 11. `debrief_cosmonauts`
Duis ac erat in felis rutrum sodales.

### 12. `stellar_archiving`
Ut efficitur, lectus in eleifend faucibus, sem justo feugiat risus, eget consequat velit libero ac felis.

**Reply by Taragolis** [reply]
There is pretty tricky to grab this information in case of dynamically generate multiple DAGs from the single module.
It also could add inconvenience for some one who doesn't required this behaviour, as well as this increase size of serialised DAG.
So better generate this documentation by your own because you know what you tried to achieve.
